### DISCUSSION TITLE
Connecting over Ethernet using SDK

### DISCUSSION THREAD

**Discussion by tcorbly** [iRobotEducation/create3_docs#440] [Questions]
### How are you connecting to your Create 3?

USB Ethernet (ROS 2)

### Computer(s) Model(s) and Operating System(s)

Raspberry Pi 3B/ Raspberry Pi OS (Bullseye)

### Which version of ROS 2 is installed on your computer?

I am unable to install ROS 2

### Which firmware version is installed on your robot?

G.5.2

### Which RMW is your robot running?

I don't know

### Is the robot connected to a network? If so, what is the network type?

Yes, home network

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

No

### What is the Adapter Board's USB/BLE Toggle currently switched to?

USB

### Describe your question.

I am attempting to connect to the Create 3 using a USB-C to Ethernet adapter and connecting over Ethernet between that and the Pi 3B. I have been able to set up a static IP in `/etc/network/interfaces` using the snippet below and ping the Create's IP with no issues. 
```
auto eth0
iface eth0 inet static
        address 192.168.186.3
        netmask 255.255.255.0
        gateway 192.168.186.255
        dns-nameservers 8.8.8.8
```

I am trying to use the [irobot-edu-python-sdk](https://github.com/iRobotEducation/irobot-edu-python-sdk/tree/main) to send commands to the create but I have not been able to get it to connect yet.

My original thought was to try and use the serial option but so far I haven't had any luck with that, which may be due to the Pi 3B only using `/dev/serial0` for serial connections rather than `tty` options.

The other option I have been attempting is to connect using a a TCP socket with the Pi as a client over port 8883 and matching it to the Backend class in the SDK. I saw something with similarities in this [community submission](https://edu.irobot.com/learning-library/intro-to-ros-2-serial-communication-with-create-3) and was using it as a reference for this method. The code below is what I have been attempting to make work for this method:

**TCP Class:**
```python
from irobot_edu_sdk.backend.backend import Backend
from irobot_edu_sdk.packet import Packet
from binascii import hexlify, unhexlify
import socket
from sys import exit
from time import sleep

class TCP(Backend):
	def __init__(self, _CREATE_IP = '192.168.186.2', _CREATE_PORT = 8883):
		self.CREATE_IP = _CREATE_IP
		self.CREATE_PORT = _CREATE_PORT
		try:
			self.client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		except socket.error:
			print('Failed to create socket')
			exit()
		print('Socket created')

	async def connect(self):
		"""Connect to robot"""
		try:
			print('Connecting...')
			self.client.connect((self.CREATE_IP,self.CREATE_PORT))
		except socket.error as msg:
			print('Failed to connect to robot')
			print(msg)
			exit()

		print('Socket connected to' + self.CREATE_IP)

	async def is_connected(self) -> bool:
		"""Returns True if robot is connected"""
		try:
			self.client.sendall(b'test')
		except socket.error:
			return False
		return True

	async def disconnect(self):
		"""Disconnect from robot"""
		self.client.close()

	async def write_packet(self, packet: Packet):
		"""Write one packet to the robot"""
		try:
			self.client.sendall(packet.to_bytes())
		except socket.error:
			print('Failed to send data')

	async def read_packet(self) -> Packet:
		"""Read one packet from the robot"""
		try:
			msg = self.client.recv(1024)
		except socket.error:
			print('Failed to read data')
		
		return Packet.from_bytes(msg)
```
**Test Code:**
```python
async def main():
    print("Hello World!")
    client = TCP()
    await client.connect()

if __name__ == "__main__":
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main())
```

I receive the error below when trying to connect:
```
Hello World!
Socket created
Connecting...
Failed to connect to robot
[Errno 111] Connection refused
```
My question essentially is, how can connect over Ethernet while also using the SDK? We are avoiding using ROS because of how large it is and the SD card we use for our controllers is only 8 GB.

**Comment by shamlian** [comment]
The python web playground is for interfacing with Create 3 and Root robots over their BLE connection. If you want to use Python and Ethernet, you'll need to use ROS 2 and rclpy. Create 3 does not present a serial link over USB; the code you linked out to is taking advantage of forwarding a USB serial device over a wireless network in order to communicate with that device remotely.
